### ISSUE TITLE
slow parallelisation

### ISSUE THREAD

**Issue by gobbios** [AlexMielke1988/NetFACS#8]
After integrating the cpp functions into `netfacs()` I realised that the parallelisation actually seems to slow down the function compared to when using a plain `lapply()` loop. So I reorganised `netfacs()` in a way that allows choosing between parallel `pblapply()` and plain `lapply()`. We can also (for now) switch between using the 'old' function behaviour (using `arules` and `vegan`) and the new cpp functions. Two quick benchmarks show that `lapply()` is faster regardless of whether we use cpp or not, with the combination of `lapply()` and cpp being fastest. This seems quite odd to me. Can you confirm that @AlexMielke1988?

```
data(emotions_set)

foo <- function(rand = 10, xparallel = TRUE, usecpp = FALSE, combsize = 2) {
  netfacs(data = emotions_set[[1]],
          condition = emotions_set[[2]]$emotion,
          test.condition = 'anger',
          ran.trials = rand, 
          combination.size = combsize,
          use_cpp = usecpp, 
          use_parallel = xparallel)
}

library(microbenchmark)
combsize = 2
rand = 100
microbenchmark(foo(rand = rand, xparallel = TRUE, usecpp = FALSE, combsize = combsize), 
               foo(rand = rand, xparallel = FALSE, usecpp = FALSE, combsize = combsize),
               foo(rand = rand, xparallel = TRUE, usecpp = TRUE, combsize = combsize), 
               foo(rand = rand, xparallel = FALSE, usecpp = TRUE, combsize = combsize),
               times = 5)

#                                                                     expr  min   lq mean median   uq  max neval
#  foo(rand = rand, xparallel = TRUE, usecpp = FALSE, combsize = combsize) 6689 6835 7186   7210 7428 7767     5
# foo(rand = rand, xparallel = FALSE, usecpp = FALSE, combsize = combsize) 2477 2540 2831   2575 2640 3922     5
#   foo(rand = rand, xparallel = TRUE, usecpp = TRUE, combsize = combsize) 4384 4545 4583   4586 4618 4783     5
#  foo(rand = rand, xparallel = FALSE, usecpp = TRUE, combsize = combsize)  353  364  424    383  436  583     5


combsize = 3
rand = 1000
microbenchmark(foo(rand = rand, xparallel = TRUE, usecpp = FALSE, combsize = combsize), 
               foo(rand = rand, xparallel = FALSE, usecpp = FALSE, combsize = combsize),
               foo(rand = rand, xparallel = TRUE, usecpp = TRUE, combsize = combsize), 
               foo(rand = rand, xparallel = FALSE, usecpp = TRUE, combsize = combsize),
               times = 5)

#                                                                     expr   min    lq  mean median    uq   max neval
#  foo(rand = rand, xparallel = TRUE, usecpp = FALSE, combsize = combsize) 37.60 38.31 39.14  38.49 40.27 41.01     5
# foo(rand = rand, xparallel = FALSE, usecpp = FALSE, combsize = combsize) 62.90 63.04 63.35  63.10 63.38 64.34     5
#   foo(rand = rand, xparallel = TRUE, usecpp = TRUE, combsize = combsize)  8.74  8.75  9.03   8.86  9.12  9.70     5
#  foo(rand = rand, xparallel = FALSE, usecpp = TRUE, combsize = combsize)  5.59  5.64  5.89   5.72  5.72  6.78     5
```

To summarise, on the bright side, cpp increases speed, but strangely enough the parallelisation seems to slow it down.

**Comment by gobbios** [comment]
I played with it some more, and the problem seems to be with `pblapply()`, at least on my computer. I tried another approach for parallelisation, using `mclapply()`. Code and results are below. In a nutshell: `mclapply()` does indeed speed up things compared to `pblapply()` and plain `lapply()`. My only worry is that this might be a Mac versus Windows thing. Can you check that you get similar results under Windows, @AlexMielke1988? On the other hand, it's also possible that the way `pblapply()` is set up causes the slow-down. But I don't know enough about it.

```
library(parallel)
library(doParallel)
library(pbapply)
library(arules)
library(microbenchmark)

internal_foo <- function(x) apriori(x)

some_foo <- function(x) {
  x <- list(letters[1:4], letters[3:2], letters[6:3])
  internal_foo(x)
}

# number of cores used
nc <- 3

foo_pblapply <- function(nc) {
  mycluster <- makeCluster(spec = nc, type = "PSOCK") 
  clusterExport(cl = mycluster, 
                c("some_foo", "internal_foo", "apriori"),
                envir = environment()) 
  registerDoParallel(mycluster)
  temp <- pblapply(1:100, cl = mycluster, FUN = some_foo)
  stopCluster(mycluster)
  temp
}

foo_lapply <- function(nc) {
  lapply(1:100, some_foo)
}

foo_mclapply <- function(nc) {
  mclapply(1:100, some_foo, mc.cores = nc)
}

x1 <- foo_pblapply(3)
x2 <- foo_lapply(3)
x3 <- foo_mclapply(3)

identical(x1, x2)
identical(x1, x3)

microbenchmark(foo_pblapply(nc = 3),
               foo_lapply(nc = 3),
               foo_mclapply(nc = 3),
               times = 5)

#                expr       min        lq      mean    median        uq       max neval
# foo_pblapply(nc = 3) 3695.7199 3761.4445 4350.7288 4381.4454 4905.2245 5009.8097     5
#   foo_lapply(nc = 3) 2065.5497 2088.6104 2309.0641 2100.8316 2283.3558 3006.9731     5
# foo_mclapply(nc = 3)  827.7017  846.8538  887.9378  907.2318  920.2183  937.6836     5
```

**Comment by AlexMielke1988** [comment]
Similar response as you for the first simulation: Cpp non-parallel fastest.

# foo(rand = rand, xparallel = TRUE, usecpp = FALSE)  41.54 41.86 46.18  43.14 43.47 60.87     5
# foo(rand = rand, xparallel = FALSE, usecpp = FALSE  21.76 25.72 29.71  28.14 34.80 38.12     5
# foo(rand = rand, xparallel = TRUE, usecpp = TRUE)   31.13 32.02 36.48  32.59 40.06 46.59     5
# foo(rand = rand, xparallel = FALSE, usecpp = TRUE)    2.82   2.99   3.34    3.27   3.47   4.15     5

I continue with 2 combinations maximum to make things faster, but 1000 randomisations. Here, for the non-cpp, parallelisation is faster, but for the cpp it is not. Cpp outperforms parallel non-cpp.
foo(rand = rand, xparallel = TRUE, usecpp = FALSE)  28.60 28.65 29.44  28.69 29.87 31.04     3
foo(rand = rand, xparallel = FALSE, usecpp = FALSE) 37.76 37.77 37.97  37.78 38.08 38.37     3
foo(rand = rand, xparallel = TRUE, usecpp = TRUE)   10.86 10.93 11.20  11.00 11.37 11.74     3
foo(rand = rand, xparallel = FALSE, usecpp = TRUE)    4.94   4.94   4.97    4.95   4.99   5.03     3 


mclapply creates problems on my Windows machine. If I run the code only 100 times, on my machine the different parallelisations and lapply are pretty much the same. However, when I run it 1000 times, 'foreach' really outperforms the lapply and pblapply:

                 expr             min     lq   mean  median      uq   max neval
 foo_pblapply(nc = 3) 18.4 18.99   21.47   21.85    23.12   24.9     5
   foo_lapply(nc = 3)   35.8  35.89  38.41   38.10    40.72    41.5     5
  foo_foreach(nc = 3)   5.1    5.78    5.93     5.78      6.22      6.8     5

This is the foo_foreach script, maybe try it out too and see how it performs:

foo_foreach <- function(nc) {
  mycluster <- makeCluster(spec = nc, type = "PSOCK") 
  clusterExport(cl = mycluster, 
                c("some_foo", "internal_foo", "apriori"),
                envir = environment()) 
  registerDoParallel(mycluster)
  foreach(i = 1:1000) %dopar% some_foo
}

**Comment by gobbios** [comment]
Great. Although I think there is a tiny typo in `foo_foreach` which biases its performance to its advantage. Code is below. Comparing the four possibilities it seems that `mclapply` is fastest after all. So I'll start by modifying `netfacs` to use `mclapply`. If you're okay with that just close the issue.


```
library(parallel)
library(doParallel)
library(pbapply)
library(arules)
library(microbenchmark)
library(foreach)

internal_foo <- function(x) apriori(x)

some_foo <- function(x) {
  x <- list(letters[1:4], letters[3:2], letters[6:3])
  internal_foo(x)
}

# number of cores used
nc <- 3

foo_pblapply <- function(nc, runs = 1000) {
  mycluster <- makeCluster(spec = nc, type = "PSOCK") 
  clusterExport(cl = mycluster, 
                c("some_foo", "internal_foo", "apriori"),
                envir = environment()) 
  registerDoParallel(mycluster)
  temp <- pblapply(1:runs, cl = mycluster, FUN = some_foo)
  stopCluster(mycluster)
  temp
}

foo_lapply <- function(nc, runs = 1000) {
  lapply(1:runs, some_foo)
}

foo_mclapply <- function(nc, runs = 1000) {
  mclapply(1:runs, some_foo, mc.cores = nc)
}

foo_foreach <- function(nc, runs = 1000) {
  mycluster <- makeCluster(spec = nc, type = "PSOCK")
  clusterExport(cl = mycluster,
                c("some_foo", "internal_foo", "apriori"),
                envir = environment())
  registerDoParallel(mycluster)
  temp <- foreach(i = 1:runs) %dopar% some_foo()
  stopCluster(mycluster)
  temp
}

x1 <- foo_pblapply(3)
x2 <- foo_lapply(3)
x3 <- foo_mclapply(3)
x4 <- foo_foreach(3)

identical(x1, x2)
identical(x1, x3)
identical(x1, x4)

microbenchmark(foo_pblapply(nc = 3, runs = 1000),
               foo_lapply(nc = 3, runs = 1000),
               foo_mclapply(nc = 3, runs = 1000),
               foo_foreach(nc = 3, runs = 1000),
               times = 5)

#                              expr       min        lq      mean    median        uq       max neval
# foo_pblapply(nc = 3, runs = 1000)  8.333979  8.486216  8.836360  8.559027  9.056805  9.745773     5
#   foo_lapply(nc = 3, runs = 1000) 13.692893 13.730115 14.332491 13.780983 14.382938 16.075523     5
# foo_mclapply(nc = 3, runs = 1000)  4.123995  4.129013  4.969626  4.487248  5.050524  7.057348     5
#  foo_foreach(nc = 3, runs = 1000)  7.078375  7.481898  7.866698  7.489201  8.550297  8.733721     5

#                             expr       min        lq      mean   median        uq       max neval
# foo_pblapply(nc = 3, runs = 100) 2993.3227 3200.8978 3342.2804 3210.685 3553.1106 3753.3862     5
#   foo_lapply(nc = 3, runs = 100) 1201.9637 1309.5953 1369.4599 1410.082 1435.3506 1490.3080     5
# foo_mclapply(nc = 3, runs = 100)  697.1521  708.8909  824.3077  823.572  899.8323  992.0914     5
#  foo_foreach(nc = 3, runs = 100) 2374.1210 2745.3176 2722.3718 2757.461 2843.6573 2891.3023     5
```

**Comment by AlexMielke1988** [comment]
Sorry, I think that bit got lost; mclapply does not work on my machine, tells me that it is not made for Windows or something. So that is not an option. 

Maybe we can build in a decision that lapply is used if people have fewer than 500 randomisations, and it switches to parallel if they have made than 500 randomisations?

**Comment by AlexMielke1988** [comment]
With foreach, that decision would be easy, because we simply have to have an if... then that swaps foreach() %dopar% and %do%

**Comment by gobbios** [comment]
Ah okay, I misunderstood you then. If `mclapply` doesn't work at all, then sure it's not an option as a general solution. Let me think about the implementation then. It would be nice to let Mac users take advantage of `mcapply`, but I wonder how that works with CRAN policies. 
In any case, I wouldn't make the switch between `lapply` and `foreach` only dependent on the number of randomisations because other things like number of events, combination size and number of units will affect how more efficient parallelisation is compared to sequential. My preference would be to keep the decision as a function argument.

**Comment by AlexMielke1988** [comment]
we can do a function argument that is use_parallel with TRUE, FALSE, or 'mac'

**Comment by gobbios** [comment]
`netfacs()` has now these two arguments: `use_parallel = TRUE` and `parallel_safe = TRUE`. `use_parallel` decides whether to use parallelisation. `parallel_safe` determines whether to use the 'safe' way, i.e. `parallel::parLapply`, which should work on all platforms. If setting `parallel_safe = FALSE`, the function will try to use `mcapply` on Unix based systems (Mac, Linux). With the default settings it should work on all platforms. And if you use a Mac, you'd need to proactively set `parallel_safe = FALSE` to get the advantage of `mcapply` (which seems to be quite a bit faster than `parLapply`, `foreach` and `pblapply` (at least under the conditions I tested).
